doc/manpages/unlznt1.1
.TH RVA2OFS 1
.SH NAME
unlznt1 - uncompress LZNT1 data

.SH SYNOPSIS
.B unlznt1
.IR file

.SH DESCRIPTION
unlznt1 uncompress LZNT1 data files (eventually found in malware)
.PP
\&\fIfile\fR is any readable file.

.SH OPTIONS

.TP
.BR \-v ", " \-\-version
show program version and exit.

.TP
.BR \-\-help
show help.

.SH EXAMPLES
Uncompress a file named \fBdumped\fP:
.IP
$ unlznt1 dumped

.SH REPORTING BUGS
Please, check the latest development code and report at https://github.com/merces/pev/issues

.SH COPYRIGHT
Copyright © 2015 pev authors. License GPLv2+: GNU GPL version 2 or later <https://www.gnu.org/licenses/gpl-2.0.txt>.
This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law.

CHANGELOG
pev - the PE file analysis toolkit - ChangeLog

Legend:
+ Added feature
* Improved/changed feature
- Bug fixed
! Known issue / missing feature

pev 0.70 - ?
 ! Missing full/English documentation.
 ! Missing valid XML and HTML output formats.
 ! pestr: no support for --net option with unicode strings.
 ! pestr: unable to handle too big strings.
 * libpe: rewritten, now using mmap. (Jardel Weyrich).
 * pestr: added countries domains suffixes.
 * readpe and peres: output enhancements (Jardel Weyrich).
 + pehash: sections and headers hash calculation (Jardel Weyrich).
 + pehash: ssdeep fuzzy hash calculation.
 + pehash: support for new digest hashes like sha512, ripemd160 and more.
 + peres: added new tool to analyze/extract PE resources (Marcelo Fleury).
 + pescan: cpl malware detection.
 + pescan: undocumented anti-disassembly fpu trick detection.
 - readpe can't show functions exported by ID only.
 - readpe: fixed subsystem types (Dmitry Mostovenko).

pev 0.60 - October 31, 2012
 ! Missing full/English documentation.
 ! Missing valid XML and HTML output formats.
 ! pestr: no support for --net option with unicode strings.
 ! readpe can't show functions exported by ID only.
 * pedis: -F/--function option replaced by -r/--rva.
 + added manpages for all tools.
 + pedis: added -m/--mode option to set disassembly mode (16, 32 or 64-bit).
 + pedis: added -n option to limit number of disassembled instructions.
 + pedis: added options to disassembly entrypoint and raw fille offset.
 + pedis: disassemble bytes number specified by -n option.
 + pehash: new tool to calculate PE file hashes (Jan Seidl).
 + pepack: added PEiD signature search (Rodrigo Escobar).
 + pescan: added -f/--format option to format output.
 + pescan: added section, imagebase and timestamp analysis.
 + readpe: added --exports option to show exported functions.
 - pedis: fixed address representation in calls and jump instructions.

pev 0.50 - June 25, 2012
 ! Missing documentation.
 ! Missing valid XML and HTML output formats.
 ! pedis shows arguments of jumps and calls as relative positions.
 * Improved pev tools Makefile (Gabriel Barbosa).
 * MEW packer detection in packid (Rodrigo Rubira).
 * pev now is a collection of binaries and a library to work with PE executables.
 + Output in monospaced text and csv in most programs.
 + Released libpe 1.0 to support our programs.
 + libpe: xmalloc trick and fixes (Rodrigo Rubira).
 + pedis: disassemble functions and sections (Tiago Zaniquelli).
 + pepack: detect fake EP (Wagner Barongello).
 + pescan: new tool to search for suspicious things in PE files including TLS callbacks.
 + pesec: find security features in PE files.
 + readpe can now show imported functions with --imports or -i switch.
 + readpe: show PE headers and sections information (most of obsolete pev binary).
 + rva2ofs and ofs2rva: convert from rva to raw file offset and vice-versa.
 - Fixed erroneuous ord numbers in functions imported without name.
 - Fixed two bugs with fake TLS callbacks in petls (thanks to Qualys guys for reporting).

pev 0.40 - August 7, 2011
 * Compatible with PE/COFF specification v8.2.
 * Date format in COFF header similar to RFC 2822.
 * Improved function to get machine type (Gabriel Duarte).
 + ASLR and DEP identification.
 + Added "-r" option to show resource items at first level.
 + Added TLS callback functions detection in every PE section.
 + Added more human-readable fields, like subsystem and section characteristics.
 + PE32+ support. Now pev can handle 64-bits executables.
 + Variable data directories support (no more fixed in 16).

pev 0.31 - May 11, 2011
 + Added characteristics flags in COFF output.
 + Added human-readable machine types in COFF output.
 - Fixed warning with Linux 32-bits boxes when compiling.
 - Fixed compilation in OS X. (Gustavo Roberto).

pev 0.30 - February 20, 2011
 * Improved memory use.
 * Now pev shows the Product Version with option "-p".
 + Added option "-a" to show all information.
 + Added option "-c" to show the COFF header.
 + Added option "-d" to show the DOS header.
 + Added option "-o" to show the Option (PE) header.
 + Added option "-s" to show executable sections.
 
pev 0.22 - January 9, 2011
 ! Does not support PE32+ files.
 ! Plans to read more PE informations.
 * Improved Makefile.
 + Added manpage.

pev 0.2 - December 26, 2011
 * Improved search algorithm.
 - Fixed bug compiling in MS-Windows platform.

pev 0.1 - December 12, 2010
   Initial release. (Eduardo Fernandes, Fernando Mercês, Francivan Bezerra and Thiago Moraes).


src/unlznt1.c
/*
	pev - the PE file analyzer toolkit

	unlznt1.c - Decompress LZNT1 data

	Copyright (C) 2015 pev authors

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <getopt.h>

#define PROGRAM "unlznt1"

static void usage(void)
{
	printf("Usage: %s FILE\n"
		"Decompress LZNT1 data\n"
		"\nExample: %s file.lznt1 > file\n"
		"\nOptions:\n"
		" -v, --version                          show version and exit\n"
		" --help                                 show this help and exit\n",
		PROGRAM, PROGRAM);
}

static void *parse_options(int argc, char *argv[])
{
	/* Parameters for getopt_long() function */
	static const char short_options[] = "o:v";

	static const struct option long_options[] = {
		{ "help",			no_argument,		NULL,	 1  },
		{ "version",		no_argument,		NULL,	'v' },
		{ NULL,				0,					NULL, 	 0  }
	};

	int c, ind;

	while ((c = getopt_long(argc, argv, short_options, long_options, &ind)))
	{
		if (c < 0)
			break;

		switch (c)
		{
			case 1: // --help option
				usage();
				exit(EXIT_SUCCESS);
			default:
				fprintf(stderr, "%s: try '--help' for more information\n", PROGRAM);
				exit(EXIT_FAILURE);
		}
	}
}

int main(int argc, char *argv[])
{
	if (argc < 2) {
		usage();
		exit(EXIT_FAILURE);
	}

	parse_options(argc, argv);

	const char *path = argv[argc-1];

	ssize_t aux = 0;
	int i;
	FILE *fp = fopen(path, "rb");
	fseek(fp, 0, SEEK_END);
	size_t siz = ftell(fp);
	fseek(fp, 0, SEEK_SET);
	
	char *data = malloc(siz);
	if (!data) {
		fprintf(stderr, "fatal: memory exhausted (malloc of %zu bytes)\n", siz);
		return EXIT_FAILURE;
	}

	if (fread(data, siz, 1, fp) != 1) {
		fprintf(stderr, "fatal: error when reading file content\n");
		return EXIT_FAILURE;
	}

	fclose(fp);
	aux = lznt1_decompress(data, siz, NULL);

	if (aux < 1) {
		fprintf(stderr, "fatal: data probably not LZNT1 compressed\n");
		return EXIT_FAILURE;
	}
	
	char *u = malloc(aux);
	if (!u) {
		fprintf(stderr, "fatal: buffer memory exhausted (malloc of %zu bytes)\n", aux);
		return EXIT_FAILURE;
	}

	memset(u, 0, aux);
	lznt1_decompress(data, siz, u);
	free(data);

	for (i=0; i<aux; i++)
		putchar(u[i]);

	free(u);
	
	return EXIT_SUCCESS;
}